File: backend/app/schemas/admin_reportes.py
from pydantic import BaseModel
from datetime import date, datetime
from typing import List, Optional


# ===== AUSENCIAS =====

class AusenciaDetalle(BaseModel):
    id: int
    colaborador_id: int
    nombre_colaborador: str
    sector_id: int
    fecha: date
    motivo: str
    created_at: datetime

    class Config:
        from_attributes = True


class RankingAusencia(BaseModel):
    colaborador_id: int
    nombre: str
    sector_id: int
    cantidad_ausencias: int
    porcentaje_semana: float

    class Config:
        from_attributes = True


class ResumenAusencias(BaseModel):
    ranking: List[RankingAusencia]
    detalle: List[AusenciaDetalle]
    total_registros: int
    periodo: dict  # {fecha_inicio, fecha_fin}


# ===== FRANJAS =====

class DistribucionFranjaItem(BaseModel):
    fecha: date
    franja_id: int
    franja_nombre: str
    hora_inicio: str
    hora_fin: str
    asignados: int
    ausentes: int
    capacidad: int
    disponibles_backlog: int

    class Config:
        from_attributes = True


class CumplimientoPreferencia(BaseModel):
    colaborador_id: int
    nombre: str
    sector_id: int
    franja_preferida_id: Optional[int]
    franja_preferida_nombre: Optional[str]
    total_asignaciones: int
    asignaciones_en_preferencia: int
    porcentaje_cumplimiento: float

    class Config:
        from_attributes = True


class CoberturaPorFranja(BaseModel):
    franja_id: int
    franja_nombre: str
    hora_inicio: str
    hora_fin: str
    orden: int
    ocupacion_promedio: float
    capacidad_promedio: float
    porcentaje_cobertura: float

    class Config:
        from_attributes = True


class ResumenFranjas(BaseModel):
    distribucion: List[DistribucionFranjaItem]
    cumplimiento_preferencias: List[CumplimientoPreferencia]
    cobertura_real: List[CoberturaPorFranja]
    periodo: dict


# ===== SWAPS =====

class SwapDetalle(BaseModel):
    id: int
    solicitante_id: int
    solicitante_nombre: str
    receptor_id: int
    receptor_nombre: str
    fecha: date
    franja_origen: str
    franja_receptor: str
    estado: str
    motivo_rechazo: Optional[str] = None
    created_at: datetime
    dias_antiguedad: int

    class Config:
        from_attributes = True


class RankingSwapColaborador(BaseModel):
    colaborador_id: int
    nombre: str
    sector_id: int
    swaps_pendientes: int
    swaps_aceptados: int
    swaps_rechazados: int
    total: int

    class Config:
        from_attributes = True


class EstadisticasSwaps(BaseModel):
    total_pendientes: int
    total_aceptados: int
    total_rechazados: int
    total_general: int


class ResumenSwaps(BaseModel):
    ranking: List[RankingSwapColaborador]
    detalle: List[SwapDetalle]
    estadisticas: EstadisticasSwaps
    periodo: dict

File: frontend/src/api/admin_reportes.ts
import client from './client'

// Ausencias
export interface AusenciaDetalle {
  id: number
  colaborador_id: number
  nombre_colaborador: string
  sector_id: number
  fecha: string
  motivo: string
  created_at: string
}

export interface RankingAusencia {
  colaborador_id: number
  nombre: string
  sector_id: number
  cantidad_ausencias: number
  porcentaje_semana: number
}

export interface ResumenAusencias {
  ranking: RankingAusencia[]
  detalle: AusenciaDetalle[]
  total_registros: number
  periodo: { fecha_inicio: string; fecha_fin: string }
}

// Franjas
export interface DistribucionFranjaItem {
  fecha: string
  franja_id: number
  franja_nombre: string
  hora_inicio: string
  hora_fin: string
  asignados: number
  ausentes: number
  capacidad: number
  disponibles_backlog: number
}

export interface CumplimientoPreferencia {
  colaborador_id: number
  nombre: string
  sector_id: number
  franja_preferida_id: number | null
  franja_preferida_nombre: string | null
  total_asignaciones: number
  asignaciones_en_preferencia: number
  porcentaje_cumplimiento: number
}

export interface CoberturaPorFranja {
  franja_id: number
  franja_nombre: string
  hora_inicio: string
  hora_fin: string
  orden: number
  ocupacion_promedio: number
  capacidad_promedio: number
  porcentaje_cobertura: number
}

export interface ResumenFranjas {
  distribucion: DistribucionFranjaItem[]
  cumplimiento_preferencias: CumplimientoPreferencia[]
  cobertura_real: CoberturaPorFranja[]
  periodo: { fecha_inicio: string; fecha_fin: string }
}

// Swaps
export interface SwapDetalle {
  id: number
  solicitante_id: number
  solicitante_nombre: string
  receptor_id: number
  receptor_nombre: string
  fecha: string
  franja_origen: string
  franja_receptor: string
  estado: string
  motivo_rechazo: string | null
  created_at: string
  dias_antiguedad: number
}

export interface RankingSwapColaborador {
  colaborador_id: number
  nombre: string
  sector_id: number
  swaps_pendientes: number
  swaps_aceptados: number
  swaps_rechazados: number
  total: number
}

export interface EstadisticasSwaps {
  total_pendientes: number
  total_aceptados: number
  total_rechazados: number
  total_general: number
}

export interface ResumenSwaps {
  ranking: RankingSwapColaborador[]
  detalle: SwapDetalle[]
  estadisticas: EstadisticasSwaps
  periodo: { fecha_inicio: string; fecha_fin: string }
}

export const adminReportesApi = {
  getAusencias: (fechaInicio: string, fechaFin: string, sectorId?: number) => {
    const params = new URLSearchParams({
      fecha_inicio: fechaInicio,
      fecha_fin: fechaFin,
    })
    if (sectorId) params.append('sector_id', sectorId.toString())
    return client.get<ResumenAusencias>(`/admin/reportes/ausencias?${params.toString()}`)
  },

  getFranjas: (fechaInicio: string, fechaFin: string, sectorId?: number) => {
    const params = new URLSearchParams({
      fecha_inicio: fechaInicio,
      fecha_fin: fechaFin,
    })
    if (sectorId) params.append('sector_id', sectorId.toString())
    return client.get<ResumenFranjas>(`/admin/reportes/franjas?${params.toString()}`)
  },

  getSwaps: (fechaInicio: string, fechaFin: string, sectorId?: number) => {
    const params = new URLSearchParams({
      fecha_inicio: fechaInicio,
      fecha_fin: fechaFin,
    })
    if (sectorId) params.append('sector_id', sectorId.toString())
    return client.get<ResumenSwaps>(`/admin/reportes/swaps?${params.toString()}`)
  },

  exportCsv: (seccion: 'ausencias' | 'franjas' | 'swaps', fechaInicio: string, fechaFin: string, sectorId?: number) => {
    const params = new URLSearchParams({
      seccion,
      fecha_inicio: fechaInicio,
      fecha_fin: fechaFin,
    })
    if (sectorId) params.append('sector_id', sectorId.toString())
    return client.get(`/admin/reportes/export/csv?${params.toString()}`, {
      responseType: 'blob',
    })
  },

  exportPdf: (seccion: 'ausencias' | 'franjas' | 'swaps', fechaInicio: string, fechaFin: string, sectorId?: number) => {
    const params = new URLSearchParams({
      seccion,
      fecha_inicio: fechaInicio,
      fecha_fin: fechaFin,
    })
    if (sectorId) params.append('sector_id', sectorId.toString())
    return client.get(`/admin/reportes/export/pdf?${params.toString()}`, {
      responseType: 'blob',
    })
  },
}
